cs/test/core/DeserializerControlsTests.cs

```diff
@@ -0,0 +1,153 @@
+                var xmlWriter = new SimpleXmlWriter(XmlWriter.Create(xmlString));
+
+                Serialize.To(xmlWriter, obj);
+                xmlWriter.Flush();
+
+                var reader = new SimpleXmlReader(XmlReader.Create(new StringReader(xmlString.ToString())));
+                T obj2 = new Deserializer<SimpleXmlReader>(typeof(T)).Deserialize<T>(reader);
+
+                Assert.True(Comparer.Equal<T>(obj, obj2));
+            }
+        }
+
+        void RoundtripTests()
+        {
+            var containers = new SimpleContainers();
+            for (int i = 0; i < 30; i++)
+            {
+                containers.strings.Add(i.ToString());
+                containers.numbers.Add(i, i.ToString());
+            }
+
+            RoundtripObject<SimpleContainers>(containers);
+
+            var blobs = new StructWithBlobs();
+            blobs.b = new ArraySegment<byte>(blobData, 0, 30);
+            blobs.lb.Add(new ArraySegment<byte>(blobData, 0, 30));
+            blobs.lb.Add(new ArraySegment<byte>(blobData, 1, 30));
+
+            RoundtripObject<StructWithBlobs>(blobs);
+        }
+
+        [Test]
+        public void DeserializerControls_InvalidArguments()
+        {
+            Assert.Throws<ArgumentOutOfRangeException>(() => DeserializerControls.Active.MaxPreallocatedContainerElements = -1);
+            Assert.Throws<ArgumentOutOfRangeException>(() => DeserializerControls.Active.MaxPreallocatedBlobBytes = -1);
+        }
+
+        [Test]
+        public void DeserializerControls_InvalidData()
+        {
+            DeserializerControls.Active.MaxPreallocatedContainerElements = 20;
+            DeserializerControls.Active.MaxPreallocatedBlobBytes = 20;
+
+            // These tests MUST return EndOfStreamException
+            Assert.Throws<EndOfStreamException>(() => DeserializeBadData<SimpleContainers>());
+            Assert.Throws<EndOfStreamException>(() => DeserializeBadData<Lists>());
+            Assert.Throws<EndOfStreamException>(() => DeserializeBadData<Vectors>());
+            Assert.Throws<EndOfStreamException>(() => DeserializeBadData<Sets>());
+            Assert.Throws<EndOfStreamException>(() => DeserializeBadData<Maps>());
+            Assert.Throws<EndOfStreamException>(() => DeserializeBadData<StructWithBlobs>());
+            Assert.Throws<EndOfStreamException>(() => DeserializeBadData<StructWithBonded>());
+            Assert.Throws<EndOfStreamException>(() => DeserializeBadData<StructWithByteLists>());
+
+            DeserializerControls.Active.MaxPreallocatedContainerElements = Int32.MaxValue;
+            DeserializerControls.Active.MaxPreallocatedBlobBytes = Int32.MaxValue;
+
+            // These tests may return EndOfStreamException or OutOfMemoryException depending
+            // environment -- either is acceptable
+            Assert.Throws(Is.InstanceOf<EndOfStreamException>().Or.InstanceOf<OutOfMemoryException>(), () => DeserializeBadData<SimpleContainers>());
```

Comment:
Minor: can `Is.InstanceOf<EndOfStreamException>().Or.InstanceOf<OutOfMemoryException>()` be lifted to a variable to minimize repetition and line length?

```
var acceptableException = Is.InstanceOf<EndOfStreamException>().Or.InstanceOf<OutOfMemoryException>();
Assert.Throws(acceptableException, () => DeserializeBadData<SimpleContainers>());
```